plutus-pab/src/Plutus/PAB/Core/ContractInstance/STM.hs

```diff
@@ -0,0 +1,120 @@
+{-
+
+Types and functions for contract instances that communicate with the outside
+world via STM. See note [Contract instance thread model].
+
+-}
+module Plutus.PAB.Core.ContractInstance.STM(
+    OpenEndpoint(..)
+    , BlockchainEnv(..)
+    , InstanceState(..)
+    ) where
+
+import           Control.Concurrent.STM (TMVar, TVar)
+import           Data.Aeson             (Value)
+import           Data.Map               (Map)
+import           Data.Set               (Set)
+import           Ledger                 (Address, Slot, TxId)
+import           Ledger.AddressMap      (AddressMap)
+
+{- Note [Contract instance thread model]
+
+In the PAB we run each contract instance in its own thread, following the
+design principles for concurrency described in [1].
+
+As a result
+* We use STM for concurrency
+* We try to avoid queues and use TVars where possible
+
+Contract instances can make requests to services that are managed by the PAB,
+such as the wallet backend, the chain index and the node. From the contract's
+POV we assume that these requests are responded to instantaneously. To handle
+these requests the PAB uses the
+'Wallet.Emulator.MultiAgent.EmulatedWalletEffects' list of effects.
+
+In addition to making requests to PAB services, contract instances can wait for
+events to happen. The events that can be waited upon are produced by the
+blockchain (transactions added to the ledger, new slots starting) and by
+external clients of the PAB including end-users (calling contract endpoints).
+To handle this kind of waiting the PAB uses STM and the types defined in this
+module.
+
+# QoS
+
+One of the main goals of the queueless STM design is to avoid a degradation
+of the quality of service (QoS) when the system is operating at capacity. This
+comes at a price: When the system is under pressure, some updates may be
+dropped. In practice this a result of the behaviour of STM's 'retry' primitive,
+which only guarantees to retry at some point (not immediately) after a variable
+has changed. So if the variable changes again before the retry happens, the
+intermediate state is not visible.
```

Comment:
It should also be noted that we can assume the PAB is running in a friendly environment, ie. it is much less likely to be the target of DoS attacks than the node is. In fact the node is the most likely source of increased traffic for us. So as long as we can deal with the incoming transactions from the node fast enough, we can be a little more relaxed about QoS in other places.